python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -549,11 +647,69 @@ class AssetBackfillIterationResult(NamedTuple):
     backfill_data: AssetBackfillData
 
 
+def get_asset_backfill_iteration_materialized_partitions(
+    backfill_id: str,
+    asset_backfill_data: AssetBackfillData,
+    asset_graph: ExternalAssetGraph,
+    instance_queryer: CachingInstanceQueryer,
+) -> Iterable[Optional[AssetGraphSubset]]:
+    """Returns the partitions that have been materialized by the backfill.
+
+    This function is a generator so we can return control to the daemon and let it heartbeat
+    during expensive operations.
+    """
+    recently_materialized_asset_partitions = AssetGraphSubset(asset_graph)
+    for asset_key in asset_backfill_data.target_subset.asset_keys:
+        records = instance_queryer.get_materialization_records(
+            asset_key=asset_key, after_cursor=asset_backfill_data.latest_storage_id
+        )
+        records_in_backfill = [
+            record
+            for record in records
+            if instance_queryer.run_has_tag(
+                run_id=record.run_id, tag_key=BACKFILL_ID_TAG, tag_value=backfill_id
+            )
+        ]
+        recently_materialized_asset_partitions |= {
+            AssetKeyPartitionKey(asset_key, record.partition_key) for record in records_in_backfill
+        }
+
+        yield None
+
+    updated_materialized_subset = (
+        asset_backfill_data.materialized_subset | recently_materialized_asset_partitions
+    )
+
+    yield updated_materialized_subset
+
+
+def get_asset_backfill_iteration_failed_and_downstream_partitions(
```

Comment:
Calling this `_get_failed_and_downstream_asset_partitions` would make it more consistent with `_get_failed_asset_partitions`. Or we could change the name of that function. Either way, I don't think we need the word "iteration" in the title - in theory this function could be used outside the context of an iteration.